### PR TITLE
feat: add DNS resolution for preferred chunk GET nodes (PE-8488)

### PR DESCRIPTION
**Pull request by djwhitt**
## Summary
- Implements DNS resolution for preferred chunk GET and POST nodes to improve performance
- Adds data-1 through data-12.arweave.xyz as default GET nodes (http port 1984)
- Adds tip-2 through tip-4.arweave.xyz as default POST nodes (http port 1984)
- Reduces DNS lookup overhead during chunk operations by pre-resolving and caching IPs

### DNS Resolution Module
- Added `DnsResolver` class with IPv4/IPv6 support and caching
- Resolves hostnames to IP addresses on startup
- Maintains resolution cache to avoid repeated DNS lookups
- Supports periodic re-resolution to handle IP changes
- Uses `net.isIP()` for IP address detection
- Properly handles IPv6 addresses with bracket notation

### Configuration Updates
- Added default preferred chunk GET nodes (data-1 through data-12.arweave.xyz:1984)
- Added default preferred chunk POST nodes (tip-2 through tip-4.arweave.xyz:1984)
- Added `PREFERRED_CHUNK_NODE_DNS_RESOLUTION_INTERVAL_SECONDS` configuration (default: 3600 seconds)
- Added URL validation for both GET and POST node configurations
- Users can override defaults via environment variables

### ArweaveCompositeClient Integration
- Integrated DNS resolver with proper encapsulation
- Added `initializeDnsResolution()` method for startup resolution (idempotent)
- Added `stopDnsResolution()` method for cleanup
- Tracks both original and resolved URLs for GET and POST nodes
- Updates weighted peer lists when DNS resolution changes
- Handles both IPv4 and IPv6 addresses with proper URL formatting
- Consolidated DNS resolution and peer updates into single timer to eliminate duplication
- Preferred POST peers maintain high weight regardless of failures

### Periodic Re-resolution
- Configurable interval for re-resolving DNS (default: 1 hour)
- Automatically updates resolved IPs when DNS changes
- Preserves non-preferred peers while updating preferred ones
- Runs in background without blocking operations
- Change detection with snapshot comparison

## Technical Details
- DNS resolution happens on startup via `arweaveClient.initializeDnsResolution()`
- Resolution failures are non-fatal and logged as warnings
- Resolved IPs are used for all chunk GET and POST operations
- Falls back gracefully to original URLs if DNS resolution fails
- Preserves ports and paths when constructing resolved URLs
- Handles IPv6 addresses with proper bracket notation
- Resolution runs asynchronously without blocking startup
- Single timer manages both DNS resolution and peer updates

## Testing
- ✅ All unit tests pass (`node --import ./register.js --test src/lib/dns-resolver.test.ts`)
- ✅ Lint checks pass (`yarn lint:check`)
- ✅ Build successful (`yarn build`)
- ✅ DNS resolution handles IPv4, IPv6, ports, and paths correctly
- ✅ Periodic re-resolution detects and updates changed IPs
- ✅ Comprehensive test coverage for all DNS resolution scenarios
- ✅ Preferred POST peer weight management maintains high priority

### Environment Variables
```bash
# Preferred chunk GET nodes (defaults to data-1 through data-12.arweave.xyz:1984)
PREFERRED_CHUNK_GET_NODE_URLS=http://custom1.example.com:1984,http://custom2.example.com:1984

# Preferred chunk POST nodes (defaults to tip-2 through tip-4.arweave.xyz:1984)
PREFERRED_CHUNK_POST_NODE_URLS=http://custom1.example.com:1984,http://custom2.example.com:1984

# DNS resolution interval in seconds (0 to disable, default: 3600)
PREFERRED_CHUNK_NODE_DNS_RESOLUTION_INTERVAL_SECONDS=3600
```

## Performance Impact
- Reduces DNS lookup overhead by caching resolutions
- Improves chunk retrieval and submission latency
- Minimal memory overhead for resolution cache
- Background re-resolution doesn't impact performance

## Recent Updates
- Fixed IPv6 bracket handling in Node.js URL API
- Added idempotent DNS initialization to prevent timer orphaning
- Improved error handling with early URL parsing
- Fixed DNS change detection with proper snapshot comparison
- Added DNS shutdown to system shutdown sequence
- Consolidated duplicate timers into single resolution timer
- Renamed configuration to `PREFERRED_CHUNK_NODE_DNS_RESOLUTION_INTERVAL_SECONDS`

## Jira Ticket
[PE-8488](https://ardrive.atlassian.net/browse/PE-8488)

🤖 Generated with [Claude Code](https://claude.ai/code)

[PE-8488]: https://ardrive.atlassian.net/browse/PE-8488?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/arweave/composite-client.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@ import { context, trace, Span } from '@opentelemetry/api';
 import { ReadThroughPromiseCache } from '@ardrive/ardrive-promise-cache';
 
 import { FailureSimulator } from '../lib/chaos.js';
+import { DnsResolver } from '../lib/dns-resolver.js';
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/arweave/composite-client.ts:23]: WAT? We use `.js` everywhere in this project. Why would it be different here?

File: src/app.ts [hunk 2]
```diff
@@ -20,6 +20,9 @@ import { apolloServer } from './routes/graphql/index.js';
 import { openApiRouter } from './routes/openapi.js';
 import * as system from './system.js';
 
+// Initialize DNS resolution for preferred chunk GET nodes
+await system.arweaveClient.initializeDnsResolution();
+
 system.arweaveClient.refreshPeers();
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/app.ts]: We don't want to await `refreshPeers`. That would be too slow. Catching DNS errors might be okay though. I'll get Claude to take a look.

File: src/arweave/composite-client.ts [hunk 3]
```diff
@@ -328,21 +341,217 @@
...
+        'Initialized preferred chunk GET URLs with DNS resolution',
+        {
+          originalUrls: this.preferredChunkGetUrls.length,
+          resolvedUrls: this.resolvedChunkGetUrls.length,
+          usingDefaults: this.preferredChunkGetUrls.some(
+            (url) => url.includes('data-') && url.includes('.arweave.xyz'),
+          ),
+        },
+      );
+    }
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/arweave/composite-client.ts]: I think this might be handled elsewhere already. I'll get Claude to take look though.

File: src/arweave/composite-client.ts [hunk 4]
```diff
@@ -328,21 +341,217 @@
...
 
-  private initializeChunkPostPeers(): void {
-    // Initialize weightedPostChunkPeers with preferred URLs at high weight
-    this.weightedPostChunkPeers = this.preferredChunkPostUrls.map(
-      (peerUrl) => ({
+  async initializeDnsResolution(): Promise<void> {
+    if (!this.dnsResolver) {
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/arweave/composite-client.ts:380]: Yeah, this seems good. I'll get Claude to pull this in.

File: src/arweave/composite-client.ts [hunk 5]
```diff
@@ -328,21 +341,217 @@ export class ArweaveCompositeClient
...
-    this.weightedGetChunkPeers = this.preferredChunkGetUrls.map((peerUrl) => ({
+    // Initialize weightedGetChunkPeers with resolved URLs at high weight
+    this.weightedGetChunkPeers = this.resolvedChunkGetUrls.map((peerUrl) => ({
       id: peerUrl,
       weight: 100, // High weight for preferred chunk GET URLs
     }));
+
+    // Log URL resolution for debugging
+    if (this.dnsResolver && this.preferredChunkGetUrls.length > 0) {
+      this.log.info(
```
Comment (djwhitt) [inline src/arweave/composite-client.ts]: This should be debug logging.

File: src/arweave/composite-client.ts [hunk 6]
```diff
@@ -328,21 +341,217 @@ export class ArweaveCompositeClient
...
+      return;
+    }
+
+    // Update GET URLs
+    const newResolvedGetUrls = this.dnsResolver.getResolvedUrlStrings(
+      this.preferredChunkGetUrls,
+    );
+
+    // Check if GET URLs have changed
+    const getUrlsChanged =
```
Comment (djwhitt) [inline src/arweave/composite-client.ts]: This seems very clunky. Is there a more efficient way to handle this?

File: src/arweave/composite-client.ts [hunk 7]
```diff
@@ -328,21 +341,217 @@ export class ArweaveCompositeClient
...
+      this.weightedGetChunkPeers = [...preferredPeers, ...nonPreferredPeers];
+    }
+
+    // Update POST URLs
+    const newResolvedPostUrls = this.dnsResolver.getResolvedUrlStrings(
+      this.preferredChunkPostUrls,
     );
+
+    // Check if POST URLs have changed
+    const postUrlsChanged =
```
Comment (djwhitt) [inline src/arweave/composite-client.ts]: Same comment here about clunkiness.

Comment (djwhitt) [reply src/arweave/composite-client.ts]: Also there's some duplication across what we're doing here and the handling of the GET URLs. Could we consolidate that somehow?

File: src/lib/dns-resolver.ts [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+   * Get all resolved URLs
+   */
+  getAllResolvedUrls(): ResolvedUrl[] {
+    return Array.from(this.resolvedUrls.values());
+  }
+
+  /**
+   * Start periodic DNS re-resolution
+   */
+  startPeriodicResolution(urls: string[], intervalMs: number): void {
```
Comment (djwhitt) [inline src/lib/dns-resolver.ts]: Are we actually using this? Seems like we might be duplicating some logic that's now in the composite client.

**Comment by djwhitt** [comment]
## ✅ Addressed CodeRabbit Feedback

The following issues have been resolved in commit 710c2a1:

1. **Fixed TypeScript build error** - The unused 'index' parameter was already removed from `src/arweave/composite-client.ts:345`

2. **Added DNS initialization error handling** - DNS resolution failures are now non-fatal during startup. The app will continue with original URLs if DNS is temporarily unavailable (`src/app.ts:24-30`)

Both critical issues identified by CodeRabbit have been addressed and the code passes all lint checks.

**Comment by djwhitt** [comment]
## ✅ All CodeRabbit Feedback Addressed

Thanks for the thorough review @coderabbitai! I've addressed all the feedback in commits 710c2a1 and f4b1fa0:

### Issues Fixed:

1. **DNS initialization error handling** (710c2a1)
   - Added try-catch to prevent DNS failures from blocking app startup
   - App continues with original URLs if DNS resolution fails

2. **Improved DNS resolver implementation** (f4b1fa0)
   - ✅ Replaced regex with `net.isIP()` for robust IP detection
   - ✅ Fixed IPv6 bracket handling when setting URL.hostname
   - ✅ Improved error handling by parsing URL once before try block
   - ✅ Fixed DNS change detection by taking snapshot before updating
   - ✅ Made `initializeDnsResolution()` idempotent by clearing existing timers
   - ✅ Added DNS resolver shutdown to system shutdown sequence
   - ✅ Prevented overlapping DNS resolutions with `isResolving` flag
   - ✅ Updated test to account for default POST URLs

### Testing:
- All unit tests pass (`yarn test`)
- DNS resolver tests specifically verified
- Lint checks pass (`yarn lint:check`)

The DNS resolution system is now more robust and handles edge cases properly. The change detection will correctly log when IPs change, and the system properly cleans up on shutdown.

**Comment by djwhitt** [comment]
## ✅ All CodeRabbit Feedback Addressed

Thank you for the thorough review! I've addressed all the feedback in the latest commits:

### Critical Issues Fixed:
- ✅ **IPv6 handling**: Fixed to properly add brackets for IPv6 addresses when setting hostname
- ✅ **IP detection**: Replaced regex with `net.isIP()` for proper IP address detection
- ✅ **Error handling**: Improved with early URL parsing before try block
- ✅ **DNS change detection**: Fixed with proper snapshot comparison
- ✅ **Idempotent initialization**: `initializeDnsResolution()` now clears existing timers
- ✅ **System shutdown**: Added DNS shutdown to system cleanup sequence
- ✅ **Timer consolidation**: Eliminated duplicate timers between DNS resolver and composite client

### Minor Improvements:
- ✅ **JSDoc documentation**: Added comprehensive JSDoc comments to all DNS-related methods
- ✅ **URL deduplication**: Using `Set` to prevent over-representation in weighted peer lists
- ✅ **Log levels**: Moved verbose DNS resolution logs from info to debug level
- ✅ **Array comparison**: Replaced `JSON.stringify` with Node.js built-in `util.isDeepStrictEqual()`
- ✅ **Environment variable**: Renamed to `PREFERRED_CHUNK_NODE_DNS_RESOLUTION_INTERVAL_SECONDS` for clarity

### Technical Improvements:
- DNS resolution is now handled by a single timer in the composite client
- Preferred POST peers maintain high weight regardless of failures
- All URLs are deduplicated to prevent duplicate entries in peer lists
- Comprehensive test coverage for all scenarios

The PR is now ready for final review. All tests pass, lint checks are clean, and the code follows best practices.
